### PR TITLE
Update transaction wire format

### PR DESCRIPTION
**Pull request by CriesofCarrots**
Adds Transaction.parseWireFormat method, and updates Transaction.serialize to match optimized wire format.
Includes helper functions for encoding/decoding bytearray length.

### CODE REVIEW

**Comment by CriesofCarrots** [comment]
I'll add some tests for the encoding helpers tomorrow, but hoping to get your eyes on this in the meantime, @mvines

**Comment by mvines** [comment]
Looks great so far, LMK when you want me to review the remaining tests/etc

**Comment by CriesofCarrots** [comment]
@mvines , ready for review

File: test/shortvec-encoding.test.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  checkDecodedArray(array, 0x100);
+
+  array = [0xff, 0xff, 0x01];
+  checkDecodedArray(array, 0x7fff);
+
+  array = [0x80, 0x80, 0x80, 0x01];
+  checkDecodedArray(array, 0x200000);
+});
+
+let prevLength = 0;
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply test/shortvec-encoding.test.js]: Actually, I wrote it that way first, and eslint complained about `prevLength` being unused. eslint exception preferred?

[the remark this answers is not in the document]
Comment (mvines) [reply test/shortvec-encoding.test.js]: Weird, can I see the eslint complaint?  But yeah globals are bad mkay

[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply test/shortvec-encoding.test.js]: Written this way:
```
function checkEncodedArray(
  array: Array<number>,
  len: number,
  prevLength: number,
  addedLength: number,
  expectedArray: Array<number>,
): number {
  prevLength += addedLength;
  encodeLength(array, len);
  expect(array.length).toEqual(prevLength);
  expect(array.slice(-addedLength)).toEqual(expectedArray);
  return prevLength;
}
```

`  52:7  error  'prevLength' is assigned a value but never used  no-unused-vars`

[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply test/shortvec-encoding.test.js]: I guess I can just do the calculation outside the function.

[the remark this answers is not in the document]
Comment (mvines) [reply test/shortvec-encoding.test.js]: After `prevLength = checkEncodedArray(array, 0, prevLength, 1, [0]);`, wouldn't you want to check that prevLength is a good value?

**Comment by CriesofCarrots** [comment]
@mvines  🎾

File: src/transaction.js [hunk 2]
```diff
@@ -386,3 +415,93 @@ export class Transaction {
     return this.instructions[0].userdata;
   }
 }
+
+/**
+ * Parse a wire transaction into a Transaction object.
+ */
+export function parseWireFormat(bytes: Buffer): Transaction {
```
Comment (mvines) [inline src/transaction.js]: I'd prefer this to be a static method of the Transaction class, 
```js
class Transaction {
.
.
/**
 * Parse a wire transaction into a Transaction object.
 */
static from(buffer: Buffer): Transaction { ... }
.
.
}
```

File: src/transaction.js [hunk 3]
```diff
@@ -191,66 +207,70 @@ export class Transaction {
       instruction.keyIndices.forEach(keyIndex => invariant(keyIndex >= 0));
     });
 
-    const instructionLayout = BufferLayout.struct([
-      BufferLayout.u8('programIdIndex'),
+    let instructionCount = [];
+    shortvec.encodeLength(instructionCount, instructions.length);
+    let instructionBuffer = Buffer.alloc(512);
```
Comment (mvines) [inline src/transaction.js]: 512 => PACKET_DATA_SIZE perhaps?

File: src/transaction.js (hunk 2 above)
Comment (CriesofCarrots) [reply src/transaction.js]: Oh duh, got it.

**Comment by mvines** [comment]
🎊

**Pull request merged**